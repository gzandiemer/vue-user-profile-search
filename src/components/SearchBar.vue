<template>
  <div>
    <input
      v-model="search"
      type="text"
      placeholder="Search by name..."
      @input="updateSearch"
    >
    <select
      v-model="gender"
      @change="updateGender"
    >
      <option value="">
        All
      </option>
      <option value="male">
        Male
      </option>
      <option value="female">
        Female
      </option>
    </select>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import { useStore } from 'vuex';

export default defineComponent({
  emits: ['update-search', 'update-gender'],
  setup(_, { emit }) {
    const store = useStore();
    
    const search = computed({
      get: () => store.state.searchQuery,
      set: (value) => {
        store.commit('setSearchQuery', value);
        emit('update-search', value);
      },
    });

    const gender = computed({
      get: () => store.state.gender,
      set: (value) => {
        store.commit('setGender', value);
        emit('update-gender', value);
      },
    });

    const updateSearch = () => {
      emit('update-search', search.value);
    };

    const updateGender = () => {
      emit('update-gender', gender.value);
    };

    return {
      search,
      gender,
      updateSearch,
      updateGender
    };
  }
});
</script>

<style></style>
