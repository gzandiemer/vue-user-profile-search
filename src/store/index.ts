import { createStore } from 'vuex';
import { User } from '../types';

interface State {
  users: User[];
  page: number;
}

export default createStore({
  state: {
    users: [],
    page: 1,
  } as State,
  mutations: {
    setUsers(state: State, users: User[]) {
      state.users = state.users.concat(users);
    },
    incrementPage(state: State){
      state.page ++;
    }
  },
  actions: {
    async fetchUsers({ commit, state }) {
      const response = await fetch(`https://randomuser.me/api/?results=25&page=${state.page}`);
      const data = await response.json();
      commit('setUsers', data.results);
      commit('incrementPage');
    }
  },
  getters: {
    getUser: (state: State) => (id: string) => {
      return state.users.find((user) => user.login.uuid === id);
    }
  }
});

