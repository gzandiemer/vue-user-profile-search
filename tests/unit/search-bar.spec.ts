import { mount } from '@vue/test-utils';

import { cloneDeep } from 'lodash';
import { createStore } from 'vuex';
import { library } from '@fortawesome/fontawesome-svg-core'
import { faSearch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'

import { State } from '@/store';

import SearchBar from '@/components/SearchBar.vue';

library.add(faSearch)

interface MyStoreOptions {
  state: State;
  actions: any;
  mutations: any;
  getters: any;
}

describe('SearchBar.vue', () => {
  let state: State;
  let actions: any;
  let mutations: any;
  let getters: any;
  let localStore: any;
  let localStoreConfig: MyStoreOptions;
  let wrapper: any;

  const mockFetchUsers = jest.fn();

  beforeEach(() => {
    state = {
      users: [],
      selectedUser: null,
      searchQuery: '',
      filter: '',
      page: 1,
      gender: ''
    };

    actions = {
      fetchUsers: mockFetchUsers
    };

    mutations = {
      setUsers: jest.fn(),
      setSelectedUser: jest.fn(),
      clearSelectedUser: jest.fn(),
      setSearchQuery: jest.fn(),
      setFilter: jest.fn(),
      setGender: jest.fn(),
      incrementPage: jest.fn()
    };

    getters = {
      getUser: jest.fn()
    };

    localStoreConfig = {
      state,
      actions,
      mutations,
      getters
    };

    localStore = createStore(cloneDeep(localStoreConfig));
  });

  // const mockFontAwesome = {
  //   install(app: any) {
  //     app.component('font-awesome-icon', {
  //       template: '<div class="fa-icon"></div>'
  //     });
  //   },
  // };

  beforeEach(() => {
    wrapper = mount(SearchBar, { global: { plugins: [localStore, FontAwesomeIcon] } });
  });


  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('emits "update-search" event and updates store on input', async () => {
    const input = wrapper.find('.search-bar__input');
    input.element.value = 'test search';
    await input.trigger('input');

    expect(mutations.setSearchQuery).toHaveBeenCalled();
    expect(wrapper.emitted('update-search')).toBeTruthy();
    expect(wrapper.emitted('update-search')?.[0]).toEqual(['test search']);
  }); //PASSES

  it('emits "update-gender" event and updates store on select', async () => {
  const dropdown = wrapper.find('.search-bar__select');
  await dropdown.trigger('click'); // opens the dropdown

  await wrapper.vm.$nextTick(); // wait for the dropdown to open

  const option = wrapper.findAll('.option').filter((w: { text: () => string; }) => w.text() === 'Male')[0];
  await option.trigger('click'); // selects the "Male" option

  // Check the mutation
  expect(mutations.setGender).toHaveBeenCalled();
});

  
});
