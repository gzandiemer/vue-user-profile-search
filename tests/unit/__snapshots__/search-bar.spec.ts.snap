// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar.vue should render correctly 1`] = `
<div class="search-bar">
  <div class="search-bar__input-container"><input class="search-bar__input" type="text" placeholder="Search by name...">
    <font-awesome-icon icon="fas,search" class="search-icon"></font-awesome-icon>
  </div>
  <div class="search-bar__select">
    <div class="selected-option">All <span class="arrow-down">▾</span><span class="arrow-up" style="display: none;">▴</span></div>
    <div class="options" style="display: none;">
      <div class="option">All</div>
      <div class="option">Male</div>
      <div class="option">Female</div>
    </div>
  </div>
</div>
`;
